contracts/CrossChain.sol

```diff
@@ -322,9 +409,143 @@ contract CrossChain is System, ICrossChain, IParamSubscriber{
+    bytes calldata payload1,
+    bytes calldata proof0,
+    bytes calldata proof1
+  )
+  onlyInit
+  blockSynced(params[0])
+  blockSynced(params[1])
+  channelSupported(uint8(params[3]))
+  whenNotSuspended
+  external {
+    // the same key with different values (payloads)
+    require(keccak256(payload0) != keccak256(payload1), "same payload");
+
+    bytes memory _key;
+    uint64 _packageSequence;
+    uint8 _channelId;
+    {
+      _packageSequence = params[2];
+      _channelId = uint8(params[3]);
+      _key = generateKey(_packageSequence, _channelId);
+      bytes32 _keyHash = keccak256(_key);
+      require(!challenged[_keyHash], "already challenged");
+
+      // if succeeding in challenge
+      challenged[_keyHash] = true;
+    }
+
+    // verify payload0 + proof0
+    {
+      uint64 _height0 = params[0];
+      bytes memory _payload0 = payload0;
+      bytes memory _proof0 = proof0;
+      bytes32 _appHash0 = ILightClient(LIGHT_CLIENT_ADDR).getAppHash(_height0);
+      require(MerkleProof.validateMerkleProof(_appHash0, STORE_NAME, _key, _payload0, _proof0), "invalid merkle proof0");
+    }
+
+    // verify payload1 + proof1
+    {
+      uint64 _height1 = params[1];
+      bytes memory _payload1 = payload1;
+      bytes memory _proof1 = proof1;
+      bytes32 _appHash1 = ILightClient(LIGHT_CLIENT_ADDR).getAppHash(_height1);
+      require(MerkleProof.validateMerkleProof(_appHash1, STORE_NAME, _key, _payload1, _proof1), "invalid merkle proof1");
+    }
+
+    _suspend();
+    emit SuccessChallenge(msg.sender, _packageSequence, _channelId);
+  }
+
+  function suspend() onlyCabinet whenNotSuspended external {
+    bool isExecutable = _approveProposal(SUSPEND_PROPOSAL, EMPTY_CONTENT_HASH);
+    if (isExecutable) {
+      _suspend();
+    }
+  }
+
+  function reopen() onlyCabinet whenSuspended external {
+    bool isExecutable = _approveProposal(REOPEN_PROPOSAL, EMPTY_CONTENT_HASH);
+    if (isExecutable) {
+      isSuspended = false;
+      emit Reopened(msg.sender);
+    }
+  }
+
+  function cancelTransfer(address tokenAddr, address attacker) onlyCabinet external {
+    bytes32 _contentHash = keccak256(abi.encode(tokenAddr, attacker));
+    bool isExecutable = _approveProposal(CANCEL_TRANSFER_PROPOSAL, _contentHash);
+    if (isExecutable) {
+      ITokenHub(TOKEN_HUB_ADDR).cancelTransferIn(tokenAddr, attacker);
+    }
+  }
+
+  function _approveProposal(bytes32 _proposalNameHash, bytes32 _contentHash) internal returns (bool isExecutable) {
+    if (quorumMap[_proposalNameHash] == 0) {
+      quorumMap[SUSPEND_PROPOSAL] = INIT_SUSPEND_QUORUM;
+      quorumMap[REOPEN_PROPOSAL] = INIT_REOPEN_QUORUM;
+      quorumMap[CANCEL_TRANSFER_PROPOSAL] = INIT_CANCEL_TRANSFER_QUORUM;
+    }
+
+    EmergencyProposal storage p = emergencyProposals[_proposalNameHash];
+
+    if (block.timestamp >= p.expiredAt || p.contentHash != _contentHash) {
+      // current proposal expired / not exist or not same with the new, create a new EmergencyProposal
```

Comment:
Please comment that it is ok if there is a bad validator always cancel the previous vote.

Reply:
fixed